mir_eval/sonify.py

```diff
@@ -147,34 +154,42 @@ def __interpolator(x):
             return value
         return __interpolator
 
+    # print(f"sonify.time_frequency length {length}, n_times {n_times}, gram shape {gram.shape}, times shape {times.shape}")
+
     # Threshold the tfgram to remove non-positive values
     gram = np.maximum(gram, 0)
 
     # Pre-allocate output signal
     output = np.zeros(length)
     time_centers = np.mean(times, axis=1) * float(fs)
 
+    # Not really a true spectral energy, just a summation for optimisation.
+    spectral_sums = np.sum(gram, axis = 1)
     for n, frequency in enumerate(frequencies):
+        if spectral_sums[n] < 0.01: # TODO set threshold intelligently.
+            continue
         # Get a waveform of length samples at this frequency
         wave = _fast_synthesize(frequency)
+        # print(f"Freq {n} {frequency}Hz energy {spectral_sums[n]}, length of wave {len(wave)}")
 
         # Interpolate the values in gram over the time grid
         if len(time_centers) > 1:
             gram_interpolator = interp1d(
-                time_centers, gram[n, :],
+                time_centers, gram[n, :-1],
                 kind='linear', bounds_error=False,
                 fill_value=(gram[n, 0], gram[n, -1]))
         # If only one time point, create constant interpolator
         else:
             gram_interpolator = _const_interpolator(gram[n, 0])
 
-        # Scale each time interval by the piano roll magnitude
-        for m, (start, end) in enumerate((times * fs).astype(int)):
+        scaler = np.zeros(length)
+        # Create the time-varying scaling for each time interval by the piano roll magnitude
+        for m, (start, end) in enumerate(sample_intervals):
             # Clip the timings to make sure the indices are valid
             start, end = max(start, 0), min(end, length)
-            # add to waveform
-            output[start:end] += (
-                wave[start:end] * gram_interpolator(np.arange(start, end)))
+            scaler[start:end] = gram_interpolator(np.arange(start, end))
+        # add to waveform
+        output += wave[:length] * scaler
```

Comment:
This look incorrect - why has the output accumulation moved outside the loop?  Won't this only include the final `scaler` array?

Reply:
In the original code, `output` was only being updated over the region `start:end` within the loop, performing the multiplication with only minimal benefit from vector multiplication, needing to iterate that `len(sample_intervals)` times. This is an optimisation where the multiplication is applied across the whole vector once, after the scaling is assembled from the `gram_interpolator()` call.

Reply:
Ok, I see now.  It won't be exactly the same as the original code in case we have intervals that overlap, but that's unsupported behavior anyway so not a big deal.



> his is an optimisation where the multiplication is applied across the whole vector once, after the scaling is assembled from the `gram_interpolator()` call.

I wonder if we could push this idea further?  We can think of this as applying the interpolator to different slices of the full sample range `np.arange(length)`.  I bet we could reorganize this, if we wanted to, to only use one call to the interpolator here, and only use the loop over intervals to construct the input sample positions that will be included.  This probably isn't a worthwhile optimization, but ought to be possible.

Reply:
Yes good point. I've now simplified it to call the interpolator once per frequency bin, and indeed, it does do worthwhile optimization, in my measurements, cutting execution time in half.